### PR TITLE
NET-1420: Dataflow command

### PR DESCRIPTION
**Pull request by seanterry**
https://seaseducation.atlassian.net/browse/NET-1420
---
### Problem
As a developer, I want to express an Extract-Transform-Load (ETL) operation as an instance of a command object so that dataflows:
- Are easily expressed in code.
- Take advantage of the validation and cancellation features of commands.

### Solution
Created dataflow command that:
- Reads records from a collection of sources.
- Executes a collection of transformations on each record.
- Notifies listeners of important events, such as when the dataflow starts and finishes, as well as when each record has completed processing. 

### Additional Notes
In this generation, metadata about dataflows has been replaced with a notification mechanism.

### CODE REVIEW

File: src/Dataflows/Dataflow.Handler.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                }
+            }
+
+            private async Task NotifyRecordCompleted( Record record, CancellationToken cancellationToken )
+            {
+                foreach ( var receiver in notificationReceivers )
+                {
+                    await receiver.NotifyRecordCompleted( record, cancellationToken );
+                }
+            }
```
Comment (seanterry) [inline src/Dataflows/Dataflow.Handler.cs:66]: These notifications replace the metadata that was added to dataflows in previous generations. This allows consuming developers to decide if and how they want to process these events.

File: src/Dataflows/Record.cs [hunk 2]
```diff
@@ -43,16 +49,17 @@ public record Record
...
         {
+            Dataflow = dataflow ?? throw new ArgumentNullException( nameof( dataflow ) );
             Source = source ?? throw new ArgumentNullException( nameof( source ) );
-            Data = data == null ? throw new ArgumentNullException( nameof( data ) ) : new Dictionary<string, object>( data, comparer );
-            Origin = data.ToImmutableDictionary( comparer );
+            Data = data == null ? throw new ArgumentNullException( nameof( data ) ) : new Dictionary<string, object>( data, dataflow.FieldNameComparer );
+            Origin = data.ToImmutableDictionary( dataflow.FieldNameComparer );
             Position = position;
         }
     }
```
Comment (seanterry) [inline src/Dataflows/Record.cs:65]: Here I switched the signature for records to give all transformations access to the configuration of the entire dataflow. This streamlined testing.

File: src/Dataflows/Transformations/AggregateTransformation.Factory.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            /// <param name="dispatcher">Transformation dispatcher.</param>
+
+            public Factory( ITransformationDispatcher dispatcher )
+            {
+                this.dispatcher = dispatcher ?? throw new ArgumentNullException( nameof( dispatcher ) );
+            }
+
+            /// <summary>
+            /// Creates and aggregate transformation handler.
+            /// </summary>
```
Comment (peterbillings7) [inline src/Dataflows/Transformations/AggregateTransformation.Factory.cs:35]: There is a typo in the summary with the word 'an'

Comment (seanterry) [reply src/Dataflows/Transformations/AggregateTransformation.Factory.cs:35]: Would for a spell checker in visual studio

File: test/Helpers/FluentValidation/ValidatorHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    /// <summary>
+    /// Extension methods to simplify validator test expressions.
+    /// </summary>
+
+    public static class ValidatorHelpers
+    {
+        /// <summary>
+        /// Asserts that the specified property is invalid when it has the given value.
+        /// </summary>
+        /// <typeparam name="TModel">Type of the partent object.</typeparam>
```
Comment (peterbillings7) [inline test/Helpers/FluentValidation/ValidatorHelpers.cs:23]: There is a typo in the param description with the word 'parent'

File: test/Helpers/FluentValidation/ValidatorHelpers.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        /// <param name="property">Property expression.</param>
+        /// <param name="value">Property value that should be considered invalid.</param>
+
+        public static IEnumerable<ValidationFailure> InvalidWhen<TModel, TValue>( this IValidator<TModel> validator, Expression<Func<TModel, TValue>> property, TValue value )
+            where TModel : class, new() => validator.ShouldHaveValidationErrorFor( property, value );
+
+        /// <summary>
+        /// Asserts that the specified property is valid when it has the given value.
+        /// </summary>
+        /// <typeparam name="TModel">Type of the partent object.</typeparam>
```
Comment (peterbillings7) [inline test/Helpers/FluentValidation/ValidatorHelpers.cs:35]: Same thing with 'parent' here

**Review by peterbillings7 — APPROVED** [review]

**Pull request merged**